src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java

```diff
@@ -2089,21 +1875,90 @@ static Object tryFinally(MethodHandle target, MethodHandle cleanup, Object... av
         return r;
     }
 
+    // see varargsArray method for chaching/package-private version of this
+    private static MethodHandle makeCollector(Class<?> arrayType, int parameterCount) {
+        MethodType type = MethodType.methodType(arrayType, Collections.nCopies(parameterCount, arrayType.componentType()));
+
+        MethodHandle newArray = MethodHandles.arrayConstructor(arrayType);
+        MethodHandle storeFunc = ArrayAccessor.getAccessor(arrayType, ArrayAccess.SET);
+
+        LambdaForm form = makeCollectorForm(type.basicType(), storeFunc);
+
+        BoundMethodHandle.SpeciesData data = BoundMethodHandle.speciesData_L();
+        BoundMethodHandle mh;
+        try {
+            mh = (BoundMethodHandle) data.factory().invokeBasic(type, form, (Object) newArray);
+        } catch (Throwable ex) {
+            throw uncaughtException(ex);
+        }
+        assert(mh.type() == type);
+        return mh;
+    }
+
+    private static LambdaForm makeCollectorForm(MethodType basicType, MethodHandle storeFunc) {
+        MethodType lambdaType = basicType.invokerType();
+        int parameterCount = basicType.parameterCount();
+
+        // Only share the lambda form for empty arrays and reference types.
+        // Sharing based on the basic type alone doesn't work because
+        // we need a separate lambda form for byte/short/char/int which
+        // are all erased to int otherwise.
+        // Other caching for primitive types happens at the MethodHandle level (see varargsArray).
+        boolean isSharedLambdaForm = parameterCount == 0 || basicType.parameterType(0) == Object.class;
+        if (isSharedLambdaForm) {
+            LambdaForm lform = basicType.form().cachedLambdaForm(MethodTypeForm.LF_COLLECTOR);
```

Comment:
How does sharing work w.r.t. array store check? Unless you put `storeFunc` on the BMH, I don't see how sharing preserves proper check.

Reply:
Nevermind, it's the type of the method handle produced `makeCollector` which ensures that all the arguments have proper type.